site/frontend/src/pages/detailed-query.ts

```diff
@@ -0,0 +1,319 @@
+import {createUrlWithParams, getUrlParams} from "../utils/navigation";
+import {postMsgpack} from "../utils/requests";
+import {SELF_PROFILE_DATA_URL} from "../urls";
+
+function to_seconds(time) {
+  return time / 1000000000;
+}
+
+function fmt_delta(to, delta, is_integral_delta) {
+  let from = to - delta;
+  let pct = (to - from) / from * 100;
+  if (from == to) {
+    pct = 0;
+  }
+  let classes;
+  if (pct > 1) {
+    classes = "positive";
+  } else if (pct < -1) {
+    classes = "negative";
+  } else {
+    classes = "neutral";
+  }
+  // some arbitrary "small" value
+  // ignore this because it's not too interesting likely
+  if (Math.abs(delta) <= 0.05) {
+    classes = "neutral";
+  }
+  let text;
+  if (is_integral_delta) {
+    text = delta.toString();
+  } else {
+    text = delta.toFixed(3);
+  }
+  if (pct != Infinity && pct != -Infinity) {
+    text += `(${pct.toFixed(1)}%)`.padStart(10, " ");
+  } else {
+    text += `-`.padStart(10, " ");
+  }
+  return `<span class="${classes}" title="${from.toFixed(3)} to ${to.toFixed(3)} ≈ ${delta.toFixed(3)}">${text}</span>`;
+}
+
+function populate_data(data, state: Selector) {
+  let txt = `${state.commit.substring(0, 10)}: Self profile results for ${state.benchmark} run ${state.scenario}`;
+  if (state.base_commit) {
+    let self_href =
+      `/detailed-query.html?sort_idx=${state.sort_idx}&commit=${state.commit}&scenario=${state.scenario}&benchmark=${state.benchmark}`;
+    let base_href =
+      `/detailed-query.html?sort_idx=${state.sort_idx}&commit=${state.base_commit}&scenario=${state.scenario}&benchmark=${state.benchmark}`;
+    txt += `<br>diff vs base ${state.base_commit.substring(0, 10)}, <a href="${base_href}">query info for just base commit</a>`;
+    txt += `<br><a href="${self_href}">query info for just this commit</a>`;
+  }
+  document.querySelector("#title").innerHTML = txt;
+  let dl_url = (commit, bench, run) => {
+    return `/perf/download-raw-self-profile?commit=${commit}&benchmark=${bench}&scenario=${run}`;
+  };
+  let dl_link = (commit, bench, run) => {
```

Comment:
The only thing that really changes in most of this code is the commit to use. So this and all the other closures are always used with constant "bench" and "run" from the state, we could remove them from the closure args and refer to the state's fields for example. There are other simplifications we could do inlining some of the closures and having the different "download/view" links be the reusable abstraction instead.